services/src/main/scala/cromwell/services/metadata/impl/builder/MetadataBuilderActor.scala

```diff
@@ -488,6 +515,134 @@ class MetadataBuilderActor(readMetadataWorkerMaker: () => Props,
       buildAndStop(query, eventsList, Map.empty, target, originalRequest)
     }
 
+  def processCostResponse(id: WorkflowId,
+                          status: WorkflowState,
+                          metadataResponse: MetadataLookupResponse,
+                          target: ActorRef,
+                          originalRequest: BuildMetadataJsonAction
+  ): State = {
+
+    // Always expand subworkflows for cost
+    val subWorkflowIds = extractSubworkflowIds(metadataResponse.eventList)
+
+    if (subWorkflowIds.isEmpty)
+      // If no workflows found, just build cost data
+      buildCostAndStop(id, status.toString, metadataResponse.eventList, Map.empty, target, originalRequest)
+    else {
+      // Otherwise spin up a metadata builder actor for each sub workflow
+      subWorkflowIds foreach { subId =>
+        val subMetadataBuilder = context.actorOf(MetadataBuilderActor.props(readMetadataWorkerMaker,
+                                                                            metadataReadRowNumberSafetyThreshold,
+                                                                            isForSubworkflows = true
+                                                 ),
+                                                 uniqueActorName(subId)
+        )
+        subMetadataBuilder ! GetCost(WorkflowId.fromString(subId))
+      }
+      goto(WaitingForSubWorkflowCost) using HasReceivedEventsData(target,
+                                                                  originalRequest,
+                                                                  metadataResponse.query,
+                                                                  metadataResponse.eventList,
+                                                                  Map.empty,
+                                                                  subWorkflowIds.size
+      )
+    }
+  }
+
+  private def extractFromJsAs[A: Manifest](js: JsObject, fieldName: String): Option[A] =
+    js.fields.get(fieldName) match {
+      case Some(a: A) => Some(a)
+      case _ => None
+    }
+
+  private def computeCost(callName: String, jsVal: JsValue): ErrorOr[BigDecimal] =
+    jsVal match {
+      case jsCall: JsObject =>
+        extractFromJsAs[JsString](jsCall, CallMetadataKeys.VmStartTime) map { startTimeVal =>
+          val startTimeErrorOr = ErrorOr(OffsetDateTime.parse(startTimeVal.value))
+
+          val endTimeErrorOr =
+            extractFromJsAs[JsString](jsCall, CallMetadataKeys.VmEndTime)
+              .map(v => ErrorOr(OffsetDateTime.parse(v.value)))
+              .getOrElse(OffsetDateTime.now().validNel)
+
+          val rawCostPerHour = extractFromJsAs[JsNumber](jsCall, CallMetadataKeys.VmCostPerHour).map(_.value)
+          val costPerHourErrorOr = rawCostPerHour match {
+            case Some(c: BigDecimal) if c >= 0 => c.validNel
```

Comment:
TIL that `JsNumber` and `BigDecimal` are the same type under the covers. Was surprised that this match works!

https://www.javadoc.io/doc/io.spray/spray-json_2.9.3/latest/spray/json/JsNumber.html